### PR TITLE
UCP/PROTO/TEST: Implement rndv/get/zcopy with new protocols

### PR DESCRIPTION
**Pull request by yosefe**
# Why
Support rendezvous protocol progress with new protocols framework

# How
- Implement RNDV progress functions 
- Add handlers for RTS/ATS
- Enable protocols in rndv tests

### CODE REVIEW

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@brminich @dmitrygx can you pls take a look?

File: src/ucp/rndv/proto_rndv.inl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    void *rkey_buffer = UCS_PTR_BYTE_OFFSET(rts, hdr_len);
+
+    rts->sreq.req_id = req->send.msg_proto.sreq_id;
+    rts->sreq.ep_id  = ucp_send_request_get_ep_remote_id(req);
+    rts->size        = req->send.state.dt_iter.length;
+
+    if (req->send.state.dt_iter.type.contig.reg.md_map == 0) {
+        rts->address = 0;
+        return hdr_len;
+    } else {
```
Comment (brminich) [inline src/ucp/rndv/proto_rndv.inl:71]: minor: can avoid this else

File: src/ucp/rndv/proto_rndv.inl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        rts->address = (uintptr_t)req->send.state.dt_iter.type.contig.buffer;
+        return hdr_len + ucp_proto_request_pack_rkey(req, rkey_buffer);
+    }
```
Comment (dmitrygx) [inline src/ucp/rndv/proto_rndv.inl:74]: can be one exit point?
```
{
    ...
    size_t rkey_size;
    ...
    if (req->send.state.dt_iter.type.contig.reg.md_map == 0) {
        rts->address = 0;
        rkey_size    = 0;
    } else {
        rts->address = (uintptr_t)req->send.state.dt_iter.type.contig.buffer;
        rkey_size    = ucp_proto_request_pack_rkey(req, rkey_buffer);
    }

    return hdr_len + rkey_size;
}
```
or:
```
{
    ...
    if (req->send.state.dt_iter.type.contig.reg.md_map == 0) {
        rts->address = 0;
        return hdr_len;
    }

    rts->address = (uintptr_t)req->send.state.dt_iter.type.contig.buffer;
    return hdr_len + ucp_proto_request_pack_rkey(req, rkey_buffer);
}
```

File: src/ucp/tag/tag_rndv.c [hunk 3]
```diff
@@ -123,11 +128,44 @@ ucs_status_t ucp_tag_send_start_rndv(ucp_request_t *sreq)
...
+                                              max_rts_size, NULL);
+}
+
 static ucp_proto_t ucp_tag_rndv_proto = {
     .name       = "tag/rndv",
     .flags      = 0,
     .init       = ucp_proto_rndv_rts_init,
     .config_str = ucp_proto_rndv_ctrl_config_str,
-    .progress   = (uct_pending_callback_t)ucs_empty_function_do_assert,
+    .progress   = ucp_tag_rndv_rts_progress,
```
Comment (brminich) [inline src/ucp/tag/tag_rndv.c:169]: no need for trailing comma

File: test/gtest/ucp/test_ucp_tag_match.cc [hunk 4]
```diff
@@ -70,6 +70,11 @@ class test_ucp_tag_match : public test_ucp_tag {
         m_req_status = status;
     }
 
+    bool use_proto() const
+    {
+        return (get_variant_value() & ENABLE_PROTO) != 0;
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_match.cc:75]: minor: imo may omit `!= 0`

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/rndv/rndv_get.c [hunk 5]
```diff
@@ -45,11 +43,112 @@ ucp_proto_rndv_get_zcopy_init(const ucp_proto_init_params_t *init_params)
...
+            req, UCP_AM_ID_RNDV_ATS, ucp_proto_request_zcopy_complete_success);
+}
+
+static ucp_proto_t ucp_rndv_ats_proto = {
+    .name       = "rndv/ats",
+    .flags      = 0,
+    .init       = ucp_proto_rndv_ats_init,
+    .config_str = ucp_proto_rndv_ack_config_str,
+    .progress   = ucp_proto_rndv_ats_progress
+};
```
Comment (brminich) [inline src/ucp/rndv/rndv_get.c:153]: do we really need to define a separate proto for ack?

File: src/ucp/rndv/proto_rndv.c [hunk 6]
```diff
@@ -303,13 +335,161 @@ ucp_proto_rndv_bulk_init(const ucp_proto_multi_init_params_t *init_params)
...
+                  "%s rva 0x%" PRIx64 " rreq_id 0x%" PRIx64 " with protocol %s",
+                  ucp_operation_names[op_id], req->send.rndv.remote_address,
+                  req->send.rndv.remote_req_id,
+                  req->send.proto_config->proto->name);
+
+    ucp_request_send(req, 0);
+    return UCS_OK;
+
+err_destroy_rkey:
+    if (rkey != NULL) {
```
Comment (brminich) [inline src/ucp/rndv/proto_rndv.c:411]: seems rkey can be non NULL here if `ucp_ep_rkey_unpack` fails

File: src/ucp/rndv/proto_rndv.inl (hunk 2 above)
Comment (yosefe) [reply src/ucp/rndv/proto_rndv.inl:74]: 1st option looks good

File: src/ucp/rndv/rndv_get.c (hunk 5 above)
Comment (yosefe) [reply src/ucp/rndv/rndv_get.c:153]: currently get_zcopy requires local registration and remote key. however for 0-length messages we don't need them, on other hand it's not possible that a protocol would require different things for different message sizes. But it's possible to define a message size range a protocol can handle.

File: src/ucp/rndv/proto_rndv.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/rndv/proto_rndv.c:411]: if ucp_ep_rkey_unpack fails we should jump to `err:` which is after this label

Comment (brminich) [reply src/ucp/rndv/proto_rndv.c:411]: right, my fault

**Review by brminich — APPROVED** [review]

**Pull request merged**